google-cloud-spanner/src/main/java/com/google/cloud/spanner/BuiltInOpenTelemetryMetricsProvider.java

```diff
@@ -0,0 +1,149 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import static com.google.cloud.opentelemetry.detection.GCPPlatformDetector.SupportedPlatform.GOOGLE_KUBERNETES_ENGINE;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.CLIENT_NAME_KEY;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.CLIENT_UID_KEY;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.DIRECT_PATH_ENABLED_KEY;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.INSTANCE_CONFIG_ID_KEY;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.LOCATION_ID_KEY;
+import static com.google.cloud.spanner.BuiltInMetricsConstant.PROJECT_ID_KEY;
+
+import com.google.auth.Credentials;
+import com.google.cloud.opentelemetry.detection.AttributeKeys;
+import com.google.cloud.opentelemetry.detection.DetectedPlatform;
+import com.google.cloud.opentelemetry.detection.GCPPlatformDetector;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.sdk.OpenTelemetrySdk;
+import io.opentelemetry.sdk.metrics.SdkMeterProvider;
+import io.opentelemetry.sdk.metrics.SdkMeterProviderBuilder;
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.reflect.Method;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.UUID;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+final class BuiltInOpenTelemetryMetricsProvider {
+
+  public static BuiltInOpenTelemetryMetricsProvider INSTANCE =
+      new BuiltInOpenTelemetryMetricsProvider();
+
+  private static final Logger logger =
+      Logger.getLogger(BuiltInOpenTelemetryMetricsProvider.class.getName());
+
+  private static String taskId;
+
+  private OpenTelemetry openTelemetry;
+
+  private BuiltInOpenTelemetryMetricsProvider() {}
+
+  OpenTelemetry getOrCreateOpenTelemetry(String projectId, @Nullable Credentials credentials) {
+    try {
+      if (this.openTelemetry == null) {
+        SdkMeterProviderBuilder sdkMeterProviderBuilder = SdkMeterProvider.builder();
+        BuiltInOpenTelemetryMetricsView.registerBuiltinMetrics(
+            SpannerCloudMonitoringExporter.create(projectId, credentials), sdkMeterProviderBuilder);
+        this.openTelemetry =
+            OpenTelemetrySdk.builder().setMeterProvider(sdkMeterProviderBuilder.build()).build();
+      }
+      return this.openTelemetry;
+    } catch (IOException ex) {
+      logger.log(
+          Level.WARNING,
+          "Unable to get OpenTelemetry object for client side metrics, will skip exporting client side metrics",
+          ex);
+      return null;
+    }
+  }
+
+  Map<String, String> getClientAttributes(
```

Comment:
nit: rename this method to `createClientAttributes(..)` so it is clear that it creates a new object every time it is invoked.